File: repository/src/main/scala/by/rekuts/travelagency/repository/ReviewSpecification.scala
package by.rekuts.travelagency.repository

import java.util

import by.rekuts.travelagency.domain.Review
import javax.persistence.criteria.{CriteriaBuilder, Predicate, Root}

class ReviewSpecification extends Specification[Review] {
  def this(id: Int) {
    this()
    this.id = id
  }

  var userId: Int = _
  var tourId: Int = _

  override def getPredicates(root: Root[Review], builder: CriteriaBuilder): util.List[Predicate] = {
    var predicates = super.getPredicates(root, builder)
    if (userId != 0) {
      predicates.add(
        builder.equal(root.get("user"), userId)
      )
    }
    if (tourId != 0) {
      predicates.add(
        builder.equal(root.get("tour"), tourId)
      )
    }
    predicates
  }
}

File: repository/src/main/java/by/rekuts/travelagency/repository/ReviewSpecification.java
package by.rekuts.travelagency.repository;

import by.rekuts.travelagency.domain.Review;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.data.jpa.domain.Specification;

import javax.persistence.criteria.CriteriaBuilder;
import javax.persistence.criteria.CriteriaQuery;
import javax.persistence.criteria.Predicate;
import javax.persistence.criteria.Root;
import java.util.ArrayList;
import java.util.List;

@Getter
@Setter
@NoArgsConstructor
public class ReviewSpecification implements Specification<Review> {
    private Integer id;
    private Integer userId;
    private Integer tourId;

    public ReviewSpecification(Integer id) {
        this.id = id;
    }

    @Override
    public Predicate toPredicate(Root<Review> root, CriteriaQuery<?> query, CriteriaBuilder builder) {
        List<Predicate> predicates = new ArrayList<>();
        if (id != null) {
            predicates.add(
                    builder.equal(root.get("id"), id)
            );
        }
        if (userId != null) {
            predicates.add(
                    builder.equal(root.get("user"), userId)
            );
        }
        if (tourId != null) {
            predicates.add(
                    builder.equal(root.get("tour"), tourId)
            );
        }
        return builder.and(predicates.toArray(new Predicate[0]));
    }
}
